File: javascript/json-transform/src/functions/TransformerFunctionConcat.ts
import { asAsyncSequence, asyncSequenceOf, emptyAsyncSequence } from "@wortise/sequency";
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined } from "../JsonHelpers";
import JsonElementStreamer from "../JsonElementStreamer";

class TransformerFunctionConcat extends TransformerFunction {
  constructor() {
    super({});
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null) return null;

    return JsonElementStreamer.fromTransformedStream(
      context,
      streamer.stream().flatMap(itm => {
        if (isNullOrUndefined(itm)) {
          return emptyAsyncSequence();
        } else if (Array.isArray(itm)) {
          return asAsyncSequence(itm);
        }
        return asyncSequenceOf(itm);
      }),
    );
  }
}

export default TransformerFunctionConcat;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionConcat.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.FunctionContext;
import co.nlighten.jsontransform.functions.common.TransformerFunction;
import co.nlighten.jsontransform.JsonElementStreamer;

import java.util.stream.Stream;

public class TransformerFunctionConcat extends TransformerFunction {
    public TransformerFunctionConcat() {
        super();
    }
    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null) return null;

        var adapter = context.getAdapter();
        return JsonElementStreamer.fromTransformedStream(context, streamer.stream()
            .flatMap(itm -> {
                if (adapter.isNull(itm)) {
                    return Stream.empty();
                } else if (adapter.isJsonArray(itm)) {
                    return adapter.stream(itm);
                }
                return Stream.of(itm);
            }));
    }
}
